pkg/containerimagebuilderpusher/kaniko.go

```diff
@@ -0,0 +1,219 @@
+package containerimagebuilderpusher
+
+import (
+	"fmt"
+	"os"
+	"os/exec"
+	"path"
+	"strings"
+	"time"
+
+	"github.com/nuclio/nuclio/pkg/errors"
+
+	"github.com/nuclio/logger"
+	batch_v1 "k8s.io/api/batch/v1"
+	v1 "k8s.io/api/core/v1"
+	meta_v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+type Kaniko struct {
+	kubeClientSet        kubernetes.Interface
+	logger               logger.Logger
+	builderConfiguration *ContainerBuilderConfiguration
+}
+
+func NewKaniko(logger logger.Logger, kubeClientSet kubernetes.Interface,
+	builderConfiguration *ContainerBuilderConfiguration) (*Kaniko, error) {
+
+	if builderConfiguration == nil {
+		return nil, errors.New("Missing kaniko builder configuration")
+	}
+
+	kanikoBuilder := &Kaniko{
+		logger:               logger,
+		kubeClientSet:        kubeClientSet,
+		builderConfiguration: builderConfiguration,
+	}
+
+	return kanikoBuilder, nil
+}
+
+func (k *Kaniko) BuildAndPushContainerImage(buildOptions *BuildOptions, namespace string) error {
+	bundleFilename, assetPath, err := k.createDockerBuildBundle(buildOptions.Image, buildOptions.ContextDir, buildOptions.TempDir)
+	if err != nil {
+		return errors.Wrap(err, "Failed to create docker build bundle")
+	}
+
+	// Remove bundle file from NGINX assets once we are done
+	defer os.Remove(assetPath) // nolint: errcheck
+
+	// Generate kaniko job spec
+	kanikoJobSpec := k.getKanikoJobSpec(buildOptions, bundleFilename)
+
+	k.logger.DebugWith("About to publish kaniko job", "jobSpec", kanikoJobSpec)
+	kanikoJob, err := k.kubeClientSet.BatchV1().Jobs(namespace).Create(kanikoJobSpec)
+	if err != nil {
+		return errors.Wrap(err, "Failed to publish kaniko job")
+	}
+
+	k.logger.Debug("Waiting for kaniko to finish")
+	timeout := time.Now().Add(10 * time.Minute)
+	for time.Now().Before(timeout) {
+		runningJob, err := k.kubeClientSet.BatchV1().Jobs(namespace).
+			Get(kanikoJob.Name, meta_v1.GetOptions{IncludeUninitialized: true})
+
+		if err != nil {
+			return errors.Wrap(err, "Failed to poll kaniko job status")
+		}
+
+		if runningJob.Status.Succeeded > 0 {
+			k.logger.Debug("Kaniko job was completed successfully")
+
+			// Cleanup
+			propagationPolicy := meta_v1.DeletePropagationBackground
+			err = k.kubeClientSet.BatchV1().Jobs(namespace).Delete(kanikoJob.Name, &meta_v1.DeleteOptions{
+				PropagationPolicy: &propagationPolicy,
+			})
+			if err != nil {
+				k.logger.Error("Failed to delete Kaniko job after successful completion")
+			}
+			return nil
+		}
+		if runningJob.Status.Failed > 0 {
+			k.logger.DebugWith("Kaniko job has failed", "status", runningJob.Status)
+			return errors.New("Kaniko job has failed")
+		}
+	}
+	return errors.New("Kaniko job has timed out")
+}
+
+func (k *Kaniko) createDockerBuildBundle(image string, contextDir string, tempDir string) (string, string, error) {
```

Comment:
can that move to docker? (`containerimagebuilderpusher` `Docker`)

I believe it would be nicer to abstract all kinds of bundling from `Kaniko`

Reply:
When docker is used to build images this bundle is not needed
It's an workaround for kaniko to share docker context between dashboard and kaniko pods

Reply:
That would limit kaniko to docker images only

Reply:
Why?
BuildBundle can be anything, I renamed it to `ContainerBuildBundle`

Reply:
assumed by the function name, the compress was for docker only. in that case, after renaming it looks good to me